core/trino-main/src/test/java/io/trino/util/TestDateTimeUtils.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.util;
+
+import org.testng.annotations.Test;
+
+import java.time.DateTimeException;
+import java.util.Optional;
+
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertFalse;
+
+public class TestDateTimeUtils
+{
+    @Test
+    public void testTryParseDateYyyyMmDd()
+    {
+        // valid dates
+        Optional<Integer> days = DateTimeUtils.tryParseDateYyyyMmDd("1970-01-01");
+        assertEquals(0, days.get().intValue());
+        days = DateTimeUtils.tryParseDateYyyyMmDd("1970-02-01");
+        assertEquals(31, days.get().intValue());
+        days = DateTimeUtils.tryParseDateYyyyMmDd("1969-12-01");
+        assertEquals(-31, days.get().intValue());
+        days = DateTimeUtils.tryParseDateYyyyMmDd("2022-02-28");
+        assertEquals(19051, days.get().intValue());
+        days = DateTimeUtils.tryParseDateYyyyMmDd("0000-01-01");
+        assertEquals(-719528, days.get().intValue());
+        days = DateTimeUtils.tryParseDateYyyyMmDd("9999-12-31");
+        assertEquals(2932896, days.get().intValue());
+
+        // format invalid
+        days = DateTimeUtils.tryParseDateYyyyMmDd("1970-2-01");
+        assertFalse(days.isPresent());
+        days = DateTimeUtils.tryParseDateYyyyMmDd("1970-02-1");
+        assertFalse(days.isPresent());
+        days = DateTimeUtils.tryParseDateYyyyMmDd("-1970-02-01");
+        assertFalse(days.isPresent());
+        days = DateTimeUtils.tryParseDateYyyyMmDd("1970/02/01");
+        assertFalse(days.isPresent());
+        days = DateTimeUtils.tryParseDateYyyyMmDd("Dec 24 2022");
+        assertFalse(days.isPresent());
+
+        // format ok, but illegal value
+        assertThatThrownBy(() -> DateTimeUtils.tryParseDateYyyyMmDd("2022-02-29"))
```

Comment:
why do we throw for this case and not for others above like 'Dec 24 2022'?

Reply:
is there a case where
```
        if (monthLow < 0 || monthLow > 9) {
            return OptionalInt.empty();
        }
```
or
```
    if (value.charAt(4) != '-') {
            return OptionalInt.empty();
        }
```
would be parsed correctly by `DATE_FORMATTER.parseMillis`?
Looks to me as if there is not. If that's the case, why not throw `DateTimeException` from `tryParseDateYyyyMmDd` if it's gonna fail anyway?

Reply:
DATE_FORMATTER.parseMillis support format other than yyyy-MM-DD, e.g. `99-12-31` and `1970-2-01`.
Lets take `1970-2-01` for example, the monthLow would be the '-', and `if (monthLow < 0 || monthLow > 9)` returns true, but we can say it's a invalid format, and throw here.
So we just delegate to DATE_FORMATTER.parseMillis for format other than yyyy-MM-DD

Reply:
You do check the length at the beginning but I guess with a possible negative year this could still be confused e.g. `-1111-1-01`.
Given https://github.com/trinodb/trino/issues/10677 we actually don't need to support negative years but I think it's better to separate those changes (perf improvement and backward incompatible change).